Etcd integration with frr. (1st development cycle)

Narutobapu opened the pull request:
### Summary
Integrate etcd client with `zebra` daemon of FRR, using `etcd_c++` library as a module hooked with `frr_late_init` and `rib_update` hooks. Client will push the routes selected by zebra to the etcd server, with IP prefix as key and route node parameters as value (nexthop, metric, address family etc.). These entries are protobuf encoded, leveraging the already implemented route entries encoding by `fpm` module of FRR.

### Components
zebra.

Narutobapu committed:
Added configuration flag '--enable-etcd_wrapper' to optionally configure etcd
module. compilation changes to add etcd module and linking etcd_c++ library
with the module along with other dependent libraries(grpc++,protobuf and grpc).

Narutobapu committed:
Added etcd module support to FRR.

Narutobapu committed:
Delete libzebra_fpm.c

Renamed to zebra_fpm.c

Narutobapu commented:
Hi,
@sangjinhan @gatanu @ravikush11 
Finishing this PR will help in speeding up the integration with other modules.

shinae-woo commented on zebra/zebra_etcd.c:
```diff
@@ -0,0 +1,639 @@
+
+/*
+ *  * Interval at which we attempt to connect to the ETCD.
+ *   */
```
Could you add units for `ZETCD_CONNECT_RETRY_IVL`? I presume that it's in second?

Narutobapu replied:
Yes, It's in seconds. Will add unit information in the comment.

shinae-woo commented on zebra/zebra_etcd.c:
```diff
@@ -0,0 +1,639 @@
+ *   * ETCD messages.
+ *    */
+#define ZETCD_OBUF_SIZE (2 * ETCD_MAX_MSG_LEN)
+#define ZETCD_IBUF_SIZE (ETCD_MAX_MSG_LEN)
```
Just for curious. What's the reason for OBUF_SIZE is larger than IBUF_SIZE? for performance, or preventing some race conditions?

shinae-woo replied:
Do we use this variables?

Narutobapu replied:
At the time of implementation, It was assumed that message will also flow from `etcd server` to `zebra`, to store those messages need a buffer with size `ZETCD_IBUF_SIZE`. But it turns out that it is not the case. Its only one way from `zebra` to `etcd server`. will remove `ZETCD_IBUF_SIZE` macro.
Will use `ZETCD_OBUF_SIZE` macro instead of hard-coded value for `obuf` (i.e. `2*4096`) while declaration.

shinae-woo commented on zebra/zebra_etcd.c:
```diff
@@ -0,0 +1,639 @@
+ *    */
+typedef struct zetcd_rnodes_iter_t_ {
+    rib_tables_iter_t tables_iter;
+      route_table_iter_t iter;
```
minor: indentation

shinae-woo replied:
Can you put a comment for each iterator separately? just for easier to read.

Narutobapu replied:
Ok, will add.

shinae-woo commented on zebra/zebra_etcd.c:
```diff
@@ -0,0 +1,639 @@
+  /*
+   * URI to connect to etcd server.
+   */
+  char zetcd_server[32];
```
Maybe, the length here become too short in different environments, Maybe we can have larger like 64, if it does not harm?

Narutobapu replied:
For now, we are accepting URI from a user in the form of dotted IPv4 `A.B.C.D` and port in the range `1-65535`. thus, `32` bytes are sufficient to store URI. Maybe, later we can increase it when accepting hostname.

shinae-woo commented on zebra/zebra_etcd.c:
Using `snprintf` will be great

Narutobapu replied:
Ok.

shinae-woo commented on zebra/zebra_etcd.c:
Can we validate the argument whether the number is in a range of 0< port < 16635?

Narutobapu replied:
Validation of port number is done by `vtysh` daemon at the time of input from a user, as the port range is declared in command declaration exclusive of boundaries.
```
DEFUN(etcd_remote_ip,
       etcd_remote_ip_cmd,
       "etcd connection ip A.B.C.D port (1-65535)",
```
should it be from 0 to 16635 only?

shinae-woo replied:
Oh, it was my typo. 1-65535, looks perfect!

shinae-woo commented on zebra/zebra_etcd.c:
`snprintf`, just in case to prevent future changes by someone else introduces any problem?

Narutobapu replied:
Ok

shinae-woo commented on zebra/zebra_etcd.c:
Do we need to define route entries size as constant?
`#define ROUTER_ENTRY_SIZE (2*4096)

Narutobapu replied:
This buffer will store all the protobuf encoded route entries before sending it to etcd server. We pass next writable buffer address to API (which encode the route entries) and we dont know the size of encoded entry, that we can't allocate it dynamically.

shinae-woo commented on zebra/zebra_etcd.c:
Just curious: Why these two variables are checked by assert? Do we expect any race conditions? Or can it be just `if(.. ) {return error;}`

Narutobapu replied:
The assertion was to avoid scheduling of the task to connect to `etcd server`, if it is already another same task scheduled in queue.
Later I realized if it asserts then program aborts, which is not expected. So, yes `if (zetcd_glob_p->t_connect) { return scheduled}` will suffice.

shinae-woo commented on zebra/zebra_etcd.c:
Just curious: Why these two variables are checked by assert? Do we expect any race conditions? Or can it be just if(.. ) {return error;}

Narutobapu replied:
Agree, same as above.

shinae-woo commented on zebra/zebra_etcd.c:
Can you have a prefix for etcd keyspace here for zebra? For example, we can get a prefix string as an argument (e.g., "/zebra/india/site1" or "/zebra/us/site1"),  then use the prefix here `"/zebra/india/site1" + prefix4`. It will help to make this works compatibly with nefeli cloud.

Narutobapu replied:
Yes, That is something I planned for next cycle.
I was thinking to have user optional input for `base prefix` along with `etcd URI`  from command, if not given the default base key prefix will be used. Also I need to add prefix length to the string of key, as while deleting route entry in data-path VM we require prefix and prefix length.

shinae-woo replied:
Sounds great!

shinae-woo reviewed (commented):
Thank you for doing this! It looks great. 
Two requests I have: 1) Having a etcd prefix key will be assigned by nefeli 2) I'm a bit worry about the per entry transaction here. We may need to measure the etcd transaction performance - e.g., how much entries can be updated per second. If the performance is not enough, we need to consider batching multiple updates as a single transaction. 3) Do you plan to have a test for zebra_etcd functionality?

Narutobapu committed:
Removed unwanted macro ZETCD_IBUF_SIZE, Added comments for iterators.

Narutobapu commented:
> Thank you for doing this! It looks great.
> Two requests I have: 1) Having a etcd prefix key will be assigned by nefeli 2) I'm a bit worry about the per entry transaction here. We may need to measure the etcd transaction performance - e.g., how much entries can be updated per second. If the performance is not enough, we need to consider batching multiple updates as a single transaction. 3) Do you plan to have a test for zebra_etcd functionality?

Thanks for comments.
1. As mentioned in on of above review comments, prefix key will be user input.
2. Yes, We should process in batch. I tried to send multiple put requests in single transaction but faced an issue, So I fall back to per entry transaction. Once we we integrate all components, will check the performance.
3. Yes, We have planned for it. For now it has been tested with add and delete static route entries into control VM.

Narutobapu committed:
Following commits:

1. replaced sprintf with snprintf.
2. Removed assertion with conditional statement.

shinae-woo reviewed (approved):
LGTM. Please merge and go ahead. Your plan looks great.

BTW. Sorry for the delay. You can feel free to leave a message on slack to me reminding any stuff I need to follow up :)

Narutobapu commented:
> LGTM. Please merge and go ahead. Your plan looks great.
> 
> BTW. Sorry for the delay. You can feel free to leave a message on slack to me reminding any stuff I need to follow up :)

Ok, Sure.
